cpp/common/src/codingstandards/cpp/alertreporting/HoldsForAllInstances.qll

```diff
@@ -0,0 +1,106 @@
+/**
+ * A module for considering whether a result occurs in all instances (e.g. copies) of the code at a
+ * given location.
+ *
+ * Multiple instances of a statement at the same location can occur for two main reasons:
+ *  1. Instantiations of a template
+ *  2. Re-compilation of a file under a different context
+ * This module helps ensure that a particular condition holds for all copies of a particular logical
+ * statement. For example, this can be used to determine whether a line of code is dead in all copies
+ * of a piece of code.
+ *
+ * This module is parameterized by a set of _candidate_ statements in the program. For each candidate
+ * statement, we determine whether all other statements that occur at the same location in the
+ * program are also part of the same set, ignoring any results generated by macros.
+ *
+ * We do so by reporting a new type of result, `LogicalResultStmt`, which represents a logical result
+ * where all instances of a statement at a given location are considered to be part of the same set.
+ */
+
+import cpp
+
+/**
+ * Holds if the `Stmt` `s` is not within a macro expansion, i.e. generated by a macro, but not the
+ * outermost `Stmt` generated by the macro.
+ */
+predicate isNotWithinMacroExpansion(Stmt s) {
+  not s.isInMacroExpansion()
+  or
+  exists(MacroInvocation mi |
+    mi.getStmt() = s and
+    not exists(mi.getParentInvocation())
+  )
+}
+
+/** A candidate set of types. */
+signature class CandidateStmtSig extends Stmt;
```

Comment:
In the future, is this likely to be extended to a more general type like `Locatable`?